### PR TITLE
[jaeger-v2] add Badger storage backend integration test

### PR DESCRIPTION
**Pull request by Pushkarm029**
## Which problem is this PR solving?
- Part of #5254 

## Description of the changes
- Added integration tests for Badger.
- Modified GH Workflow to perform integration tests for jaeger-v2 also.

## How was this change tested?
- ./scripts/badger-integration-test.sh

## Checklist
- [x] I have read https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md
- [x] I have signed all commits
- [x] I have added unit tests for the new functionality
- [x] I have run lint and test steps successfully
  - for `jaeger`: `make lint test`
  - for `jaeger-ui`: `yarn lint` and `yarn test`

### CODE REVIEW

**Comment by Pushkarm029** [comment]
@james-ryans, this test is not able to receive data.
log:
```bash
2024/03/18 12:02:01 Starting load generator at 1024 items/sec.
2024/03/18 12:02:03  | Sent:       539 traces (275/sec) | Received:         0 items (0/sec)
2024/03/18 12:02:06  | Sent:     1,533 traces (309/sec) | Received:         0 items (0/sec)
2024/03/18 12:02:08 Stopped generator. Sent:     1,533 traces (219/sec)
2024/03/18 12:02:09  | Sent:     1,533 traces (192/sec) | Received:         0 items (0/sec)
2024/03/18 12:02:12  | Sent:     1,533 traces (139/sec) | Received:         0 items (0/sec)
2024/03/18 12:02:15  | Sent:     1,533 traces (109/sec) | Received:         0 items (0/sec)
2024/03/18 12:02:18  | Sent:     1,533 traces (90/sec) | Received:         0 items (0/sec)
2024/03/18 12:02:18 Time out waiting for all data items received
    test_case.go:289: Time out waiting for all data items received
```

**Comment by james-ryans** [comment]
I've just learned that Badger is an embedded local storage. In Testbed, the collector and the data receiver are both running as separated instances. That being said the data receiver doesn't have access to anything inside the collector, that's why we can't test the memory storage backend. But from my perspective, it seems your Badger config writes and stores the spans in the `/tmp/jaeger` directory. Since both the collector and data receiver are running on the same computer, the data receiver should be able to read those spans from `/tmp/jaeger` directory. I'll try to take a look at how Badger works and update to you what I can do.

Also, that means you probably don't need to have a script to run a remote storage.

**Comment by yurishkuro** [comment]
If you point two badger instances to the same file system they may run into synchronization issues. I don't think it's recommended. 

The way to test badger and memory is by using jaeger remote storage, a binary that can run one instance to which you connect as grpc storage. I believe our crossdock tests are using that.

**Comment by james-ryans** [comment]
> If you point two badger instances to the same file system they may run into synchronization issues.

I think Yuri is right, I've just tried to run the Badger integration test with `ephemeral: false` in the config to read & write from the `directory_key` and `directory_value`. But I encountered this error.
```
panic: failed to start extensions: failed to initialize badger storage badger_main: Cannot acquire directory lock on "/tmp/jaeger".  Another process is using this Badger database. error: resource temporarily unavailable
```

> The way to test badger and memory is by using jaeger remote storage

Yes, in the previous PR for gRPC integration test, I actually used memory span storage type for the remote storage. You may also follow that configuration at `grpc_config.yaml`. But for the script, we're actually trying to improve that by build a binary/container remote storage from "this" version of repository not the published image #5266.

**Comment by Pushkarm029** [comment]
Thanks for the suggestions. I am trying it out, I will update you soon.

**Comment by Pushkarm029** [comment]
Update : This method is working fine, locally :
```bash
go run ./cmd/remote-storage
export SPAN_STORAGE_TYPE=badger
export BADGER_EPHEMERAL=false
export STORAGE=badger
make jaeger-storage-integration-test
```

but replicating this is in docker using script, it throws this error : 
```
"error":"Error Creating Dir: \"/go/bin/data/keys\" error: mkdir /go/bin/data: permission denied",
```

trying to fix this error.

**Comment by yurishkuro** [comment]
You don't want /go/bin as the data dir for badger. See https://github.com/jaegertracing/jaeger/pull/5282 about permissions

**Comment by Pushkarm029** [comment]
Fixed this permission error, Still same : 
```
2024/03/20 20:56:49 Starting load generator at 1024 items/sec.
2024/03/20 20:56:51  | Sent:       460 traces (231/sec) | Received:         0 items (0/sec)
2024/03/20 20:56:54  | Sent:     1,533 traces (307/sec) | Received:         0 items (0/sec)
2024/03/20 20:56:56 Stopped generator. Sent:     1,839 traces (268/sec)
2024/03/20 20:56:57  | Sent:     1,839 traces (230/sec) | Received:         0 items (0/sec)
2024/03/20 20:57:00  | Sent:     1,839 traces (167/sec) | Received:         0 items (0/sec)
2024/03/20 20:57:03  | Sent:     1,839 traces (131/sec) | Received:         0 items (0/sec)
2024/03/20 20:57:06  | Sent:     1,839 traces (108/sec) | Received:         0 items (0/sec)
```
Any other approach or fix @yashrsharma44

**Comment by Pushkarm029** [comment]
[dbentries.txt](https://github.com/jaegertracing/jaeger/files/14696025/dbentries.txt)
I recovered the test database to verify that traces are stored in the remote badger database and tried to read key value pair. Got this result. Please check if it is the same as what we are supposed to store in it. @james-ryans

**Comment by yurishkuro** [comment]
@Pushkarm029 I doubt you can compare binary files like that, @james-ryans might have completely different binary representation because of timestamps. You can query db via jaeger-query/apiv3 JSON interface

**Comment by Pushkarm029** [comment]
image
test data is being stored in the database, so, i think there may be some problem in data receiver.

**Comment by yurishkuro** [comment]
> traces are being stored in the database, so I think there may be some problem with the data receiver.

perhaps we can add logging to the data receiver that shows which query it is executing against the backend. Then try to run the same query manually - if it returns the data it means you likely have some connectivity issue rather than data issue.

**Comment by james-ryans** [comment]
> perhaps we can add logging to the data receiver that shows which query it is executing against the backend.

Yes, I'm trying to help figure out how to add logging similar to the OTEL collector's components log to the data receiver. 

I just finished debugging the issue, it was sourced from the artificial jaeger receiver. In Badger, the `spanReader.FindTraces` query params require `StartTimeMin` and `StartTimeMax` which causes the error below. This might also happen to @akagami-harsh's PR at #5287.
```
Failed to consume traces from consumer	{"error": "stream error: rpc error: code = Unknown desc = start and end time must be set"}
```

I'll create a new PR to fix this issue and also add logging to the data receiver. Sorry for the inconvenience.

**Comment by Pushkarm029** [comment]
> Sorry for the inconvenience.
Don't feel sorry; after all, we all are here to learn❤️

**Comment by Pushkarm029** [comment]
Why closed?

**Comment by yurishkuro** [comment]
Auto-closed because the other PR had the text `Fixes #5281 issue`, need to be careful with those.

File: cmd/jaeger/internal/integration/fixtures/badger_config.yaml [hunk 1]
```diff
@@ -0,0 +1,23 @@
+service:
+  extensions: [jaeger_storage]
+  pipelines:
+    traces:
+      receivers: [otlp]
+      exporters: [jaeger_storage_exporter]
+
+extensions:
+  jaeger_storage:
+    grpc:
```
[the remark this answers is not in the document]
Comment (Pushkarm029) [reply cmd/jaeger/internal/integration/fixtures/badger_config.yaml]: done

**Comment by yurishkuro** [comment]
@Pushkarm029  do we still need this PR or is it covered by #5355 ?

**Comment by Pushkarm029** [comment]
Nope, I'm closing it now.

**Pull request closed without merging**